core/src/SYCL/Kokkos_SYCL_UniqueToken.hpp

```diff
@@ -81,51 +86,90 @@ class UniqueToken<SYCL, UniqueTokenScope::Global> {
+  size_type impl_acquire() const {
+    auto item = sycl::ext::oneapi::experimental::this_nd_item<3>();
+    std::size_t threadIdx[3] = {item.get_local_id(2), item.get_local_id(1),
+                                item.get_local_id(0)};
+    std::size_t blockIdx[3]  = {item.get_group(2), item.get_group(1),
+                               item.get_group(0)};
+    std::size_t blockDim[3] = {item.get_local_range(2), item.get_local_range(1),
+                               item.get_local_range(0)};
+
+    int idx = blockIdx[0] * (blockDim[0] * blockDim[1]) +
+              threadIdx[1] * blockDim[0] + threadIdx[0];
+    idx %= size();
+
+    while (Kokkos::atomic_compare_exchange(&m_locks(idx), 0, 1) == 1) {
+      idx += blockDim[1] * blockDim[0] + 1;
+      idx %= size();
     }
 
-    return result.first;
+    // Make sure that all writes in the previous lock owner are visible to me
+#ifdef KOKKOS_ENABLE_IMPL_DESUL_ATOMICS
+    desul::atomic_thread_fence(desul::MemoryOrderAcquire(),
+                               desul::MemoryScopeDevice());
+#else
+    Kokkos::memory_fence();
+#endif
+    return idx;
+  }
+
+ public:
+  /// \brief acquire value such that 0 <= value < size()
+  KOKKOS_INLINE_FUNCTION
+  size_type acquire() const {
+    KOKKOS_IF_ON_DEVICE(return impl_acquire();)
+    KOKKOS_IF_ON_HOST(return 0;)
   }
 
   /// \brief release an acquired value
   KOKKOS_INLINE_FUNCTION
-  void release(size_type i) const noexcept {
-    Kokkos::Impl::concurrent_bitset::release(m_buffer, i);
+  void release(size_type idx) const noexcept {
+    // Make sure my writes are visible to the next lock owner
+#ifdef KOKKOS_ENABLE_IMPL_DESUL_ATOMICS
+    desul::atomic_thread_fence(desul::MemoryOrderRelease(),
+                               desul::MemoryScopeDevice());
+#else
+    Kokkos::memory_fence();
+#endif
+    (void)Kokkos::atomic_exchange(&m_locks(idx), 0);
   }
 };
 
 template <>
 class UniqueToken<SYCL, UniqueTokenScope::Instance>
     : public UniqueToken<SYCL, UniqueTokenScope::Global> {
-  View<uint32_t*, SYCLDeviceUSMSpace> m_buffer_view;
-
  public:
-  explicit UniqueToken(execution_space const& arg = execution_space())
-      : UniqueToken<SYCL, UniqueTokenScope::Global>(arg) {}
-
-  UniqueToken(size_type max_size, execution_space const& = execution_space())
-      : m_buffer_view(
-            "UniqueToken::m_buffer_view",
-            ::Kokkos::Impl::concurrent_bitset::buffer_bound(max_size)) {
-    m_buffer = m_buffer_view.data();
-    m_count  = max_size;
-  }
+  UniqueToken()
+      : UniqueToken<SYCL, UniqueTokenScope::Global>(
+            Kokkos::Experimental::SYCL().concurrency()) {}
+
+  explicit UniqueToken(execution_space const& arg)
+      : UniqueToken<SYCL, UniqueTokenScope::Global>(
+            Kokkos::Experimental::SYCL().concurrency(), arg) {}
+
+  explicit UniqueToken(size_type max_size)
+      : UniqueToken<SYCL, UniqueTokenScope::Global>(max_size,
+                                                    execution_space()) {}
```

Comment:
I think this one should not use the default exec space, since that forwards to the async init version.

Reply:
Fixed.